tools/kmodules/lisa-in-tree/fetch_lisa_module.py

```diff
@@ -0,0 +1,77 @@
+#! /usr/bin/env python3
+
+import os
+import shutil
+import argparse
+import subprocess
+
+
+def main():
+    parser = argparse.ArgumentParser(prog='Lisa module in-tree helper',
+                                     description="""
+    Method 1 - using LISA_HOME.
+
+    If the script finds the LISA_HOME environment variable (present after sourcing the Lisa shell)
+    it will symlink the Lisa module sources from the local Lisa git checkout into the kernel tree.
+    That way updating the local Lisa tree will automatically update the module sources in the kernel
+    tree and the changes will be present after the next kernel rebuild.
+    In this case, the script only needs to be run once.
+
+    Method 2 - Sparse checkout.
+
+    If the script doesn't find LISA_HOME (e.g. building on an external server without a Lisa checkout)
+    the alternative mode is to directly clone the module sources into the tree.
+                                     """)
+    parser.add_argument('-c', '--commit', help='Lisa git reference to checkout')
+    parser.add_argument('-f', '--force', action='store_true', help='Override the module checkout if it already exists')
+    parser.add_argument('--module-kernel-path', help='Path to where the module will be stored in-kernel',
+                        default='./private/gs-google/drivers/soc/arm/vh/kernel/lisa')
+    parser.add_argument('--module-kernel-git-dir', help='Directory name to store the module git source in-kernel',
+                        default='lisa-module-git')
+    parser.add_argument('--remote', help='Git remote to pull the module from',
+                        default='git@github.com:ARM-software/lisa.git')
+    args = parser.parse_args()
+
+    # LISA_HOME is present, use the bundled module
+    try:
+        LISA_HOME = os.environ['LISA_HOME']
+
+        if args.commit:
+            parser.error("Can't use --commit when LISA_HOME is set")
+
+        # create relative symlink
+        pwd = os.getcwd()
+        os.chdir(os.path.dirname(args.module_kernel_path))
+        os.symlink(os.path.join(LISA_HOME, 'lisa/_assets/kmodules/lisa'), os.path.basename(args.module_kernel_path))
+        os.chdir(pwd)
```

Comment:
Please no unnecessary stateful code. Use `pathlib.Path.resolve()` to turn all the paths into absolute paths at the boundaries of the script. I'd generally recommend using `pathlib` now that we mandate high enough versions of Python, it's generally easier to use and some functions actually work better than their `os` counterpart.

Reply:
That's why I was directly calling into `ln` in the first place, to use the `-r` argument to create a relative symlink. If I turn all the paths absolute then the symlink relies on the entire directory structure so if the kernel directory is moved or renamed the symlink will break whereas in the relative variant it would still work fine. We can do it with absolute paths as well, it'll just be less robust that's all.

Reply:
You can still achieve the same result without being stateful. You can use Path.relative_to() to make a symlink target relative to whatever base you want:
```
from pathlib import Path

path1 = Path('./a').resolve()
path2 = Path('./a/b').resolve()
path3 = Path('./a/c').resolve()

# At this point, all the paths are absolute, and we will make a symlink `./a/c -> ./a/b`

path1.mkdir()
path2.mkdir()

target = path2.relative_to(path3.parent)
path3.symlink_to(target)
```

Reply:
yeah this works too, thanks!